subsys/net/openthread/rpc/client/ot_rpc_meshdiag.c

```diff
@@ -0,0 +1,165 @@
+	nrf_rpc_encode_uint(&ctx, (ot_rpc_res_tab_key)aIterator);
+
+	nrf_rpc_cbor_cmd_rsp_no_err(&ot_group, OT_RPC_CMD_MESH_DIAG_GET_NEXT_IP6_ADDRESS, &ctx);
+
+	error = nrf_rpc_decode_uint(&ctx);
+	if (error != OT_ERROR_NONE) {
+		goto exit;
+	}
+
+	nrf_rpc_decode_buffer(&ctx, aIp6Address->mFields.m8, OT_IP6_ADDRESS_SIZE);
+
+exit:
+	if (!nrf_rpc_decoding_done_and_check(&ot_group, &ctx)) {
+		ot_rpc_report_rsp_decoding_error(OT_RPC_CMD_MESH_DIAG_GET_NEXT_IP6_ADDRESS);
+	}
+
+	return error;
+}
+
+otError otMeshDiagGetNextChildInfo(otMeshDiagChildIterator *aIterator,
+				   otMeshDiagChildInfo *aChildInfo)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	otError error;
+	uint8_t child_flags;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 1 + sizeof(ot_rpc_res_tab_key));
+	nrf_rpc_encode_uint(&ctx, (ot_rpc_res_tab_key)aIterator);
+
+	nrf_rpc_cbor_cmd_rsp_no_err(&ot_group, OT_RPC_CMD_MESH_DIAG_GET_NEXT_CHILD_INFO, &ctx);
+
+	error = nrf_rpc_decode_uint(&ctx);
+	if (error != OT_ERROR_NONE) {
+		goto exit;
+	}
+
+	aChildInfo->mRloc16 = nrf_rpc_decode_uint(&ctx);
+	child_flags = nrf_rpc_decode_uint(&ctx);
+	aChildInfo->mLinkQuality = nrf_rpc_decode_uint(&ctx);
+
+	aChildInfo->mMode.mRxOnWhenIdle = (bool)(child_flags & BIT(0));
+	aChildInfo->mMode.mDeviceType = (bool)(child_flags & BIT(1));
+	aChildInfo->mMode.mNetworkData = (bool)(child_flags & BIT(2));
+	aChildInfo->mIsThisDevice = (bool)(child_flags & BIT(3));
+	aChildInfo->mIsBorderRouter = (bool)(child_flags & BIT(4));
+
+exit:
+	if (!nrf_rpc_decoding_done_and_check(&ot_group, &ctx)) {
+		ot_rpc_report_rsp_decoding_error(OT_RPC_CMD_MESH_DIAG_GET_NEXT_CHILD_INFO);
+	}
+
+	return error;
+}
+
+otError otMeshDiagDiscoverTopology(otInstance *aInstance, const otMeshDiagDiscoverConfig *aConfig,
+				   otMeshDiagDiscoverCallback aCallback, void *aContext)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	otError error;
+	uint8_t config = 0;
+
+	ARG_UNUSED(aInstance);
+
+	mesh_diag_discover_cb = aCallback;
+	mesh_diag_discover_cb_context = aContext;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 1);
+
+	WRITE_BIT(config, 0, aConfig->mDiscoverIp6Addresses);
+	WRITE_BIT(config, 1, aConfig->mDiscoverChildTable);
+	nrf_rpc_encode_uint(&ctx, config);
+
+	nrf_rpc_cbor_cmd_no_err(&ot_group, OT_RPC_CMD_MESH_DIAG_DISCOVER_TOPOLOGY, &ctx,
+				ot_rpc_decode_error, &error);
+
+	return error;
+}
+
+void otMeshDiagCancel(otInstance *aInstance)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+
+	ARG_UNUSED(aInstance);
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 0);
+
+	nrf_rpc_cbor_cmd_no_err(&ot_group, OT_RPC_CMD_MESH_DIAG_CANCEL, &ctx, ot_rpc_decode_void,
+				NULL);
+}
+
+static void ot_rpc_mesh_diag_discover_topology_cb(const struct nrf_rpc_group *group,
+						  struct nrf_rpc_cbor_ctx *ctx, void *handler_data)
+{
+	otError error;
+	otMeshDiagRouterInfo *router_info = NULL;
+	uint8_t router_flags;
+
+	error = nrf_rpc_decode_uint(ctx);
+
+	if (error == OT_ERROR_RESPONSE_TIMEOUT) {
+		goto exit;
+	}
+
+	router_info = malloc(sizeof(otMeshDiagRouterInfo));
```

Comment:
just curious. Is there a reason why you had to switch to malloc?

Reply:
Wanted to just call `mesh_diag_discover_cb(error, router_info, mesh_diag_discover_cb_context)`, with `router_info` as argument, regardless if it is NULL or not. using malloc here imo allowed to make the code a bit cleaner